src/Features/Core/Portable/SplitOrMergeIfStatements/Consecutive/AbstractSplitIntoConsecutiveIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractSplitIntoConsecutiveIfStatementsCodeRefactoringProvider<TExpressionSyntax>
+        : AbstractSplitIfStatementCodeRefactoringProvider<TExpressionSyntax>
+        where TExpressionSyntax : SyntaxNode
+    {
+        protected abstract int LogicalOrSyntaxKind { get; }
+        protected sealed override int LogicalExpressionSyntaxKind => LogicalOrSyntaxKind;
+
+        protected abstract bool HasElseClauses(SyntaxNode ifStatementNode);
+
+        protected abstract (SyntaxNode, SyntaxNode) SplitIfStatementIntoElseClause(
+            SyntaxNode ifStatementNode, TExpressionSyntax condition1, TExpressionSyntax condition2);
+
+        protected abstract (SyntaxNode, SyntaxNode) SplitIfStatementIntoSeparateStatements(
+            SyntaxNode ifStatementNode, TExpressionSyntax condition1, TExpressionSyntax condition2);
+
+        protected sealed override CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument)
+            => new MyCodeAction(createChangedDocument, IfKeywordText);
+
+        protected sealed override async Task<SyntaxNode> GetChangedRootAsync(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode currentIfStatement,
+            TExpressionSyntax left,
+            TExpressionSyntax right,
+            CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            var (firstIfStatement, secondIfStatement) =
+                await CanBeSeparateStatementsAsync(document, syntaxFacts, currentIfStatement, cancellationToken).ConfigureAwait(false)
+                ? SplitIfStatementIntoSeparateStatements(currentIfStatement, left, right)
+                : SplitIfStatementIntoElseClause(currentIfStatement, left, right);
```

Comment:
do we need to pass syntaxFacts since we're passing along the document?

Reply:
I left as is.